Санкт-Петербургский Государственный Политехнический Университет   
## Дисциплина: Математическое моделирование   
# **Курсовая работа**

## Введение
### Постановка задачи:
![Формулировка задания](./meta/1.png)

## Ход работы
1.	Сначала происходит инициализация исходных данных: диапазон x1 из условия и начальное приближение, полученное на основе ручного расчета.
2.	Во внешнем цикле перебираем по x1.
3.	Проверяем сходимость и ограничения:   
3.1.	сошелся ли fsolve;   
3.2.	вычисление p4 и его проверка;   
3.3.	вычисление p1 и его проверка.   
4.	Вычисляем определитель Якоби и классифицируем точки (лишняя / бифуркационная).
5.	Сохраняем результат в таблицу.
6.	Строим бифуркационную диаграмму.
7.	Выполняем проверку (блок проверки).


## Ручной расчет
### Матрица Якоби
![Матрица Якоби](./meta/2.jpg)
### Ручной расчет для x1 = 0.2
![Ручной расчет для x1 = 0.2](./meta/3.jpg)

## Таблица

In [4]:
from IPython.display import display
import pandas as pd
table = pd.read_csv("bifurcation_table.csv")
pd.set_option('display.max_rows', None)
display(table)

x1       x2        p1      p4    det_J      Статус
0    0.18   1.3244   26.5809  0.3839   0.4668  Бифуркация
1    0.19   0.7939   20.3564  0.5490   0.3070  Бифуркация
2    0.20   0.6042   17.9669  0.6965   0.1743  Бифуркация
3    0.21   0.5046   16.5561  0.8343   0.0312  Бифуркация
4    0.22   0.4435   15.5440  0.9641  -0.1261      Лишняя
5    0.23   0.4028   14.7322  1.0865  -0.2960      Лишняя
6    0.24   0.3743   14.0358  1.2018  -0.4753      Лишняя
7    0.25   0.3537   13.4135  1.3101  -0.6604      Лишняя
8    0.26   0.3387   12.8433  1.4116  -0.8482      Лишняя
9    0.27   0.3275   12.3131  1.5065  -1.0355      Лишняя
10   0.28   0.3194   11.8159  1.5950  -1.2197      Лишняя
11   0.29   0.3135   11.3472  1.6774  -1.3989      Лишняя
12   0.30   0.3095   10.9042  1.7540  -1.5712      Лишняя
13   0.31   0.3069   10.4851  1.8251  -1.7353      Лишняя
14   0.32   0.3055   10.0884  1.8910  -1.8903      Лишняя
15   0.33   0.3052    9.7129  1.9521  -2.0356      Лишняя
16   0.34   0.3056    9.3575  2.0087  -2.1706      Лишняя
17   0.35   0.3068    9.0213  2.0610  -2.2952      Лишняя
18   0.36   0.3087    8.7035  2.1094  -2.4092      Лишняя
19   0.37   0.3111    8.4031  2.1541  -2.5128      Лишняя
20   0.38   0.3140    8.1194  2.1953  -2.6060      Лишняя
21   0.39   0.3174    7.8514  2.2333  -2.6890      Лишняя
22   0.40   0.3211    7.5984  2.2683  -2.7621      Лишняя
23   0.41   0.3253    7.3596  2.3005  -2.8255      Лишняя
24   0.42   0.3298    7.1343  2.3300  -2.8797      Лишняя
25   0.43   0.3347    6.9218  2.3571  -2.9249      Лишняя
26   0.44   0.3399    6.7213  2.3819  -2.9615      Лишняя
27   0.45   0.3454    6.5323  2.4045  -2.9897      Лишняя
28   0.46   0.3511    6.3540  2.4251  -3.0100      Лишняя
29   0.47   0.3572    6.1860  2.4437  -3.0227      Лишняя
30   0.48   0.3635    6.0276  2.4605  -3.0280      Лишняя
31   0.49   0.3702    5.8784  2.4756  -3.0263      Лишняя
32   0.50   0.3771    5.7377  2.4891  -3.0179      Лишняя
33   0.51   0.3842    5.6053  2.5010  -3.0031      Лишняя
34   0.52   0.3917    5.4806  2.5115  -2.9822      Лишняя
35   0.53   0.3994    5.3632  2.5205  -2.9555      Лишняя
36   0.54   0.4074    5.2528  2.5283  -2.9232      Лишняя
37   0.55   0.4157    5.1490  2.5347  -2.8856      Лишняя
38   0.56   0.4242    5.0515  2.5400  -2.8429      Лишняя
39   0.57   0.4331    4.9599  2.5441  -2.7954      Лишняя
40   0.58   0.4423    4.8740  2.5472  -2.7433      Лишняя
41   0.59   0.4517    4.7935  2.5492  -2.6869      Лишняя
42   0.60   0.4615    4.7182  2.5502  -2.6263      Лишняя
43   0.61   0.4716    4.6478  2.5502  -2.5617      Лишняя
44   0.62   0.4820    4.5820  2.5493  -2.4935      Лишняя
45   0.63   0.4928    4.5208  2.5476  -2.4216      Лишняя
46   0.64   0.5039    4.4640  2.5450  -2.3465      Лишняя
47   0.65   0.5154    4.4112  2.5416  -2.2682      Лишняя
48   0.66   0.5273    4.3625  2.5375  -2.1868      Лишняя
49   0.67   0.5396    4.3176  2.5326  -2.1027      Лишняя
50   0.68   0.5522    4.2765  2.5270  -2.0159      Лишняя
51   0.69   0.5653    4.2389  2.5207  -1.9265      Лишняя
52   0.70   0.5789    4.2048  2.5138  -1.8348      Лишняя
53   0.71   0.5929    4.1741  2.5063  -1.7409      Лишняя
54   0.72   0.6073    4.1468  2.4982  -1.6449      Лишняя
55   0.73   0.6223    4.1226  2.4896  -1.5468      Лишняя
56   0.74   0.6378    4.1016  2.4804  -1.4470      Лишняя
57   0.75   0.6538    4.0837  2.4707  -1.3453      Лишняя
58   0.76   0.6704    4.0688  2.4605  -1.2421      Лишняя
59   0.77   0.6876    4.0569  2.4498  -1.1373      Лишняя
60   0.78   0.7054    4.0480  2.4387  -1.0310      Лишняя
61   0.79   0.7238    4.0420  2.4271  -0.9233      Лишняя
62   0.80   0.7430    4.0389  2.4151  -0.8143      Лишняя
63   0.81   0.7628    4.0387  2.4028  -0.7040      Лишняя
64   0.82   0.7834    4.0414  2.3900  -0.5925      Лишняя
65   0.83   0.8048    4.0471  2.3769  -0.4799      Лишняя
66   0.84   0.8271    4.0557  2.3634  -0.3661      Лишняя
67   0.85   0.8502    4.0673  2.3496  -0.2512      Лишняя
68   

## Бифуркационная диаграмма
![Бифуркационная диаграмма](./hopf_bifurcation.png)

## Блок проверки
Блок проверки для  x1 = 0.2 в конце программы подтверждает корректность найденной стационарной точки и условий бифуркации Хопфа. Он использует fsolve с функцией trace_eq и начальным приближением  x2 = 0.6 , чтобы найти  x2 , при которомb Tr(J) = 0. Затем вычисляются  p1  и  p4 , а функции dx1_dt и dx2_dt проверяют, что правые части уравнений системы близки к нулю (стационарность). Функции trace_J и det_J подтверждают, что Tr(J) ~= 0 и det(J) > 0. Результаты выводятся в консоль, показывая значения и их соответствие условиям бифуркации.

In [6]:
%run main.py


Проверка для x1 = 0.20:
x2 = 0.604188
p1 = 17.966904
p4 = 0.696534
dx1/dt = 0.000000e+00 (должно быть близко к 0)
dx2/dt = 0.000000e+00 (должно быть близко к 0)
Trace(J) = 5.551115e-17 (должно быть близко к 0)
Det(J) = 0.174291 (должно быть > 0)


## Выводы по работе:
В рамках данной работы были определены и проанализированы точки бифуркации Андронова–Хопфа для заданной системы дифференциальных уравнений, а также визуализирована их зависимость в параметрической плоскости (p4, p1). Проведённый анализ позволяет выделить следующие ключевые результаты.   
   
1. Фильтрация точек: Значительная часть значений x1 в диапазоне [0.1, 1.4] была исключена из рассмотрения, так как уравнение Tr(J) = 0 имело комплексные корни или не удовлетворяло ограничениям (x2, p1, p4 > 0), (p4 > x1), что привело к ограничению диапазона в таблице до [0.18, 1.25]. Это согласуется с математическими особенностями системы, исключающими бифуркацию за пределами указанного интервала.   
2. Характеристика бифуркаций: Численное моделирование выявило устойчивые бифуркационные точки det(J) > 0, подтверждённые для x1 = 0.2 значениями p1 ~= 17.967, p4 ~= 0.6966. Проверка стационарности и условий бифуркации Tr(J) ~= 0, det(J) > 0 показала высокую точность вычислений, а визуализация подчёркивает нелинейную зависимость между p4 и p1, указывая на сложную динамику системы при изменении параметров.  
## Литература:  
1. С.М.Устинов, В.А.Зимницкий. Вычислительная математика. – СПб.: БХВ-Петербург, 2009. – 336с. – (Учебное пособие.)
2. Форсайт Дж., Малькольм М., Моулер К. Машинные методы математических вычислений. – М.: Мир, 1980. –  280с.
